cpp/src/reductions/scan/ewm.cu

```diff
@@ -0,0 +1,377 @@
+    encountered, that means that the "previous" value is downweighted by a
+    factor (for each missing value). For example this would y_2 be for one NULL:
+    data = {x_0, NULL, x_1},
+    y_2 = (1 - alpha)**2 x_0 + alpha * x_2 / (alpha + (1-alpha)**2)
+
+    As such, the pairs must be updated before summing like the adjusted case to
+    properly downweight the previous values. But now but we also need to compute
+    the normalization factors and divide the results into them at the end.
+
+    */
+
+    rmm::device_uvector<cudf::size_type> nullcnt = null_roll_up(input, stream);
+    pair_beta_adjust(input, pairs, nullcnt, stream);
+
+    rmm::device_uvector<T> nullcnt_factor(nullcnt.size(), stream, mr);
+
+    thrust::transform(rmm::exec_policy(stream),
+                      nullcnt.begin(),
+                      nullcnt.end(),
+                      nullcnt_factor.begin(),
+                      [=] __device__(T exponent) -> T {
+                        // ex: 2 -> alpha + (1  - alpha)**2
+                        if (exponent != 0) {
+                          return (1.0 - beta) + pow(beta, exponent + 1);
+                        } else {
+                          return exponent;
+                        }
+                      });
+
+    auto device_view = *column_device_view::create(input);
+    auto valid_it    = detail::make_validity_iterator(device_view);
+    auto null_and_null_count =
+      thrust::make_zip_iterator(thrust::make_tuple(valid_it, nullcnt_factor.begin()));
+    thrust::transform(
+      rmm::exec_policy(stream),
+      null_and_null_count,
+      null_and_null_count + input.size(),
+      pairs.begin(),
+      pairs.begin(),
+      [=] __device__(thrust::tuple<bool, T> null_and_null_count, pair_type pair) -> pair_type {
+        bool is_valid = thrust::get<0>(null_and_null_count);
+        T factor      = thrust::get<1>(null_and_null_count);
+
+        T ci = pair.first;
+        T cj = pair.second;
+
+        if (is_valid and (factor != 0.0)) {
+          return {ci / factor, cj / factor};
+        } else {
+          return {ci, cj};
+        }
+      });
+  }
+  compute_recurrence(pairs, stream);
+  // copy the second elements to the output for now
+  thrust::transform(rmm::exec_policy(stream),
+                    pairs.begin(),
+                    pairs.end(),
+                    output.begin(),
+                    [=] __device__(pair_type pair) -> T { return pair.second; });
+  return output;
+}
+
+template <typename T>
+std::unique_ptr<column> ewma(std::unique_ptr<aggregation> const& agg,
+                             column_view const& input,
+                             rmm::cuda_stream_view stream,
+                             rmm::mr::device_memory_resource* mr)
+{
+  CUDF_EXPECTS(cudf::is_floating_point(input.type()), "Column must be floating point type");
+  bool adjust = (dynamic_cast<ewma_aggregation*>(agg.get()))->adjust;
+  T com       = (dynamic_cast<ewma_aggregation*>(agg.get()))->com;
```

Comment:
This seems a little off. are we guaranteed that the provided agg will be of type `ewma_aggregation`, or is it possible for this to be called with an erroneous type? If we're sure it's the right type we can just use a `static_cast`. If we're not, then we should verify that the result of the `dynamic_cast` is not a null pointer.

Reply:
This still needs to be addressed.